instrumentation/aws-lambda-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambda/v1_0/TracingRequestWrapperBase.java

```diff
@@ -19,7 +21,10 @@
  */
 abstract class TracingRequestWrapperBase<I, O> extends TracingRequestHandler<I, O> {
 
-  protected static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  protected static final ObjectMapper OBJECT_MAPPER =
+      new ObjectMapper()
+          .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+          .registerModule(new JodaModule());
```

Comment:
Adding `JodaModule` this way might not be a good idea. The problem here is that if the instrumented application does not contain `JodaModule` then the instrumentation can not be applied. Are you sure that all `aws-lambda-java-core` users also use `jackson-datatype-joda`?

Reply:
`aws-lambda-java-events` uses Joda: https://github.com/aws/aws-lambda-java-libs/blob/master/aws-lambda-java-events/pom.xml#L49. It uses it in `ScheduledEvent` for example. If you don't use `JodaModule` your lambda layer will fail to parse it properly as described in the bug report. Add that module shouldn't do anything apart from properly parsing classes using Joda. Joda is basically dead so hopefully AWS will release an updated version of the events with Java Time API but this project still needs to support older versions of the events library, so I'd expect it to remain.

Reply:
I think in the original PR that added the JSON deserialization I had commented on how it's a bit tricky that we aren't using AWS's same mechanism for deserialization, which presumably isn't using jackson-datatype-joda.

More worringly is that the intent was that the instrumentation wouldn't cause exceptions in the user's function if the serialization did have problems but that seems to be happening here. I think we need to fix that. Otherwise we may want to give up on handling JSON - Joda types are one corner case but it's hard to be confident we don't continue to run into more and more corner cases for this. @kubawach Do you happen to be around to provide thoughts on this?

Reply:
> we aren't using AWS's same mechanism for deserialization, which presumably isn't using jackson-datatype-joda.

AWS might not even use Jackson I presume.
>More worringly is that the intent was that the instrumentation wouldn't cause exceptions in the user's function if the serialization did have problems but that seems to be happening here. I think we need to fix that.

I agree but I think this issue needs fixing nevertheless as it's one of AWS standard events provided by their library. Not failing on missing properties is a step in this direction BTW.
> Joda types are one corner case but it's hard to be confident we don't continue to run into more and more corner cases for this.

I think the instrumentation should parse all events in `aws-lambda-java-events`, so I wouldn't call it a corner case. I added tests for 3 of them but we can add test for **all** of them to avoid surprises.

Reply:
Hey, yes, @anuraaga, you remember well latest discussion on that and general skepticism regarding wrapper's ability to deserialize a required POJO from JSON. 
My two cents here are:

1. In general the wrapper is a customer-centric "nice to have" component. Users can always use appropriate handler directly if the wrapper does not work in their particular case. POJO-handling wrapper is even more so. It has a built-in fallback mechanism - if user declares own lambda handler as <Object, Object> the wrapper will not try to map coming data. In such case it will work exactly as it worked before adding POJO mapping extension. 

2. Interestingly AWS seems to publish the mapping code on GitHub - see https://github.com/aws/aws-lambda-java-libs/blob/master/aws-lambda-java-serialization/src/main/java/com/amazonaws/services/lambda/runtime/serialization/factories/JacksonFactory.java It's even in the M2 repo (albeit not up to date with changes on GitHub - last change 7 days ago, last release a year ago). So there could be a way to either implement the mapping same way as AWS does or even re-use their code by depending on the artifact.

To sum up:
1. Adding `FAIL_ON_UNKNOWN_PROPERTIES` will certainly not hurt (AWS does the same).
2. Date/time conversion is done with reflection, we could use the same to avoid adding Joda etc (see https://github.com/aws/aws-lambda-java-libs/blob/master/aws-lambda-java-serialization/src/main/java/com/amazonaws/services/lambda/runtime/serialization/util/SerializeUtil.java)
3. Overall strategy towards this can ultimately go one of two routes (as already signed by @anuraaga):
a) try to support everything same way as AWS (which may be tricky due to version changes) for example by using AWS artifact / re-using their mapping code
b) clearly state in the docs that POJO-mapping wrapper is a `convenience solution` that may not support all cases (various types etc) in which case users should implement the mapping on the handler side (and declare it as <Object, Object>)

I *personally* lean towards option `b)` as it gives more control to users and relives the community from burden of supporting external (AWS) deserialization implementation.

Reply:
@kubawach I think having to use `<Object, Object>` is a terrible user experience. Supporting a random POJO might be a nice to have but supporting default AWS types should not. It forces you to deserialise objects in your app consuming CPU and time unnecessarily, which might be a big drawback in many cases. Ideally you should be able to auto instrument your lambda without changes in the code.

About the code you shared, that's not their actual logic. It's a serialisation logic used by `aws-lambda-java-runtime-client` which is used for containerised your lambda for development. It's not designed for standalone use. You could copy the logic but I'd expect the crude reflection to be less performant than using the Jackson module, I haven't run any performance test to validate that, but it shouldn't be hard. Also you get rid of any side effects of that reflection. I feel like a good solid implementation is worth 74 KB, but it's not my call.

About the 2 routes you mentions my personal view as a user is they are not mutually exclusive. You can provide a good support for mapping POJO, helping people instrument their code with minimal changes and impact. At the same time you can document that this is a best effort solution and how to work around the issues you might experience. Please let me know if I can help.

Reply:
@acm19 ability to use the wrapper without code changes was the primary reason to introduce the changes we're discussing right now ;) However I side with @anuraaga that we may expect more corner cases in the future if we decide we (as a community) want to allow wrapper usage in each and every case. In this respect - this does not sound viable. But it's a "strategic" side of things. In this particular case I'd:
- set the option `FAIL_ON_UNKNOWN_PROPERTIES`
- implement types-via-reflection handling as documented in the AWS code
..but I'm not an approver so it's up to @open-telemetry/java-instrumentation-approvers to decide.

Wrt reflection and performance - I meant dynamic invocation in general, not exactly reflection. Static MethodHandles are common in the instrumentation code and known to be quite fast. Good way to avoid another dependency especially in the quite tirvial case.

Reply:
> ability to use the wrapper without code changes was the primary reason to introduce the changes we're discussing right now

I think the discussion is not on the purpose but the usefulness. My point is, from the user perspective this is an essential feature. Not only for convenience but also because having to parse objects is not only an unnecessary hassle but impractical for the reasons I explained before. Again, not my call, but I think this feature should at least:

1. Address **all** events POJOs in `aws-lambda-java-events`.
2. Support edge cases as they arrive whenever sensible.

> implement types-via-reflection handling as documented in the AWS code
..but I'm not an approver so it's up to @open-telemetry/java-instrumentation-approvers to decide.

If you prefer this way, I'm happy to change the PR accordingly. I think supporting "default" events is the essential bit missing that should be added in this PR anyway.

Reply:
> Address all events POJOs in aws-lambda-java-events.
> Support edge cases as they arrive whenever sensible.

This seems to make sense to me, thanks for discussing. The more event types we can cover in unit tests the safer it'll be.

Reply:
I'll be off for a few days, I'll update the PR with additional tests when I'm back and I'll address the rest of the comments. Thanks.

Reply:
I added SNS and S3 as well.